src/queue.rs

```diff
@@ -0,0 +1,320 @@
+use crate::config::Config;
+use crate::db::Pool;
+use crate::error::Result;
+use log::error;
+
+#[derive(Debug, Eq, PartialEq, serde::Serialize)]
+pub(crate) struct QueuedCrate {
+    #[serde(skip)]
+    id: i32,
+    pub(crate) name: String,
+    pub(crate) version: String,
+    pub(crate) priority: i32,
+}
+
+#[derive(Debug)]
+pub struct BuildQueue {
+    db: Pool,
+    max_attempts: i32,
+}
+
+impl BuildQueue {
+    pub fn new(db: Pool, config: &Config) -> Self {
+        BuildQueue {
+            db,
+            max_attempts: config.build_attempts.into(),
+        }
+    }
+
+    pub fn add_crate(&self, name: &str, version: &str, priority: i32) -> Result<()> {
+        self.db.get()?.execute(
+            "INSERT INTO queue (name, version, priority) VALUES ($1, $2, $3);",
+            &[&name, &version, &priority],
+        )?;
+        Ok(())
+    }
+
+    pub(crate) fn pending_count(&self) -> Result<usize> {
+        let res = self.db.get()?.query(
+            "SELECT COUNT(*) FROM queue WHERE attempt < $1;",
+            &[&self.max_attempts],
+        )?;
+        Ok(res.get(0).get::<_, i64>(0) as usize)
+    }
+
+    pub(crate) fn prioritized_count(&self) -> Result<usize> {
+        let res = self.db.get()?.query(
+            "SELECT COUNT(*) FROM queue WHERE attempt < $1 AND priority <= 0;",
+            &[&self.max_attempts],
+        )?;
+        Ok(res.get(0).get::<_, i64>(0) as usize)
+    }
+
+    pub(crate) fn failed_count(&self) -> Result<usize> {
+        let res = self.db.get()?.query(
+            "SELECT COUNT(*) FROM queue WHERE attempt >= $1;",
+            &[&self.max_attempts],
+        )?;
+        Ok(res.get(0).get::<_, i64>(0) as usize)
+    }
+
+    pub(crate) fn queued_crates(&self) -> Result<Vec<QueuedCrate>> {
+        let query = self.db.get()?.query(
+            "SELECT id, name, version, priority
+             FROM queue
+             WHERE attempt < $1
+             ORDER BY priority ASC, attempt ASC, id ASC",
+            &[&self.max_attempts],
+        )?;
+
+        Ok(query
+            .into_iter()
+            .map(|row| QueuedCrate {
+                id: row.get("id"),
+                name: row.get("name"),
+                version: row.get("version"),
+                priority: row.get("priority"),
+            })
+            .collect())
+    }
+
+    pub(crate) fn process_next_crate(
+        &self,
+        f: impl FnOnce(&QueuedCrate) -> Result<()>,
+    ) -> Result<()> {
+        let conn = self.db.get()?;
+
+        let queued = self.queued_crates()?;
+        let to_process = match queued.get(0) {
+            Some(krate) => krate,
+            None => return Ok(()),
+        };
+
+        match f(&to_process) {
+            Ok(()) => {
+                conn.execute("DELETE FROM queue WHERE id = $1;", &[&to_process.id])?;
+                crate::web::metrics::TOTAL_BUILDS.inc();
```

Comment:
We used to increment `TOTAL_BUILDS` regardless of whether the build succeeded or not. Was this change intentional?

Reply:
Fixed.